// @alicloud/sts20150401/dist/client.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// This file is auto-generated, don't edit it
/**
 *
 */
const tea_util_1 = __importStar(require("@alicloud/tea-util")), $Util = tea_util_1;
const openapi_client_1 = __importStar(require("@alicloud/openapi-client")), $OpenApi = openapi_client_1;
const openapi_util_1 = __importDefault(require("@alicloud/openapi-util"));
const endpoint_util_1 = __importDefault(require("@alicloud/endpoint-util"));
const $tea = __importStar(require("@alicloud/tea-typescript"));
class AssumeRoleRequest extends $tea.Model {
    constructor(map) {
        super(map);
    }
    static names() {
        return {
            durationSeconds: 'DurationSeconds',
            externalId: 'ExternalId',
            policy: 'Policy',
            roleArn: 'RoleArn',
            roleSessionName: 'RoleSessionName',
        };
    }
    static types() {
        return {
            durationSeconds: 'number',
            externalId: 'string',
            policy: 'string',
            roleArn: 'string',
            roleSessionName: 'string',
        };
    }
}
exports.AssumeRoleRequest = AssumeRoleRequest;
class AssumeRoleResponseBody extends $tea.Model {
    constructor(map) {
        super(map);
    }
    static names() {
        return {
            assumedRoleUser: 'AssumedRoleUser',
            credentials: 'Credentials',
            requestId: 'RequestId',
        };
    }
    static types() {
        return {
            assumedRoleUser: AssumeRoleResponseBodyAssumedRoleUser,
            credentials: AssumeRoleResponseBodyCredentials,
            requestId: 'string',
        };
    }
}
exports.AssumeRoleResponseBody = AssumeRoleResponseBody;
class AssumeRoleResponse extends $tea.Model {
    constructor(map) {
        super(map);
    }
    static names() {
        return {
            headers: 'headers',
            statusCode: 'statusCode',
            body: 'body',
        };
    }
    static types() {
        return {
            headers: { 'type': 'map', 'keyType': 'string', 'valueType': 'string' },
            statusCode: 'number',
            body: AssumeRoleResponseBody,
        };
    }
}
exports.AssumeRoleResponse = AssumeRoleResponse;
class AssumeRoleWithOIDCRequest extends $tea.Model {
    constructor(map) {
        super(map);
    }
    static names() {
        return {
            durationSeconds: 'DurationSeconds',
            OIDCProviderArn: 'OIDCProviderArn',
            OIDCToken: 'OIDCToken',
            policy: 'Policy',
            roleArn: 'RoleArn',
            roleSessionName: 'RoleSessionName',
        };
    }
    static types() {
        return {
            durationSeconds: 'number',
            OIDCProviderArn: 'string',
            OIDCToken: 'string',
            policy: 'string',
            roleArn: 'string',
            roleSessionName: 'string',
        };
    }
}
exports.AssumeRoleWithOIDCRequest = AssumeRoleWithOIDCRequest;
class AssumeRoleWithOIDCResponseBody extends $tea.Model {
    constructor(map) {
        super(map);
    }
    static names() {
        return {
            assumedRoleUser: 'AssumedRoleUser',
            credentials: 'Credentials',
            OIDCTokenInfo: 'OIDCTokenInfo',
            requestId: 'RequestId',
        };
    }
    static types() {
        return {
            assumedRoleUser: AssumeRoleWithOIDCResponseBodyAssumedRoleUser,
            credentials: AssumeRoleWithOIDCResponseBodyCredentials,
            OIDCTokenInfo: AssumeRoleWithOIDCResponseBodyOIDCTokenInfo,
            requestId: 'string',
        };
    }
}
exports.AssumeRoleWithOIDCResponseBody = AssumeRoleWithOIDCResponseBody;
class AssumeRoleWithOIDCResponse extends $tea.Model {
    constructor(map) {
        super(map);
    }
    static names() {
        return {
            headers: 'headers',
            statusCode: 'statusCode',
            body: 'body',
        };
    }
    static types() {
        return {
            headers: { 'type': 'map', 'keyType': 'string', 'valueType': 'string' },
            statusCode: 'number',
            body: AssumeRoleWithOIDCResponseBody,
        };
    }
}
exports.AssumeRoleWithOIDCResponse = AssumeRoleWithOIDCResponse;
class AssumeRoleWithSAMLRequest extends $tea.Model {
    constructor(map) {
        super(map);
    }
    static names() {
        return {
            durationSeconds: 'DurationSeconds',
            policy: 'Policy',
            roleArn: 'RoleArn',
            SAMLAssertion: 'SAMLAssertion',
            SAMLProviderArn: 'SAMLProviderArn',
        };
    }
    static types() {
        return {
            durationSeconds: 'number',
            policy: 'string',
            roleArn: 'string',
            SAMLAssertion: 'string',
            SAMLProviderArn: 'string',
        };
    }
}
exports.AssumeRoleWithSAMLRequest = AssumeRoleWithSAMLRequest;
class AssumeRoleWithSAMLResponseBody extends $tea.Model {
    constructor(map) {
        super(map);
    }
    static names() {
        return {
            assumedRoleUser: 'AssumedRoleUser',
            credentials: 'Credentials',
            requestId: 'RequestId',
            SAMLAssertionInfo: 'SAMLAssertionInfo',
        };
    }
    static types() {
        return {
            assumedRoleUser: AssumeRoleWithSAMLResponseBodyAssumedRoleUser,
            credentials: AssumeRoleWithSAMLResponseBodyCredentials,
            requestId: 'string',
            SAMLAssertionInfo: AssumeRoleWithSAMLResponseBodySAMLAssertionInfo,
        };
    }
}
exports.AssumeRoleWithSAMLResponseBody = AssumeRoleWithSAMLResponseBody;
class AssumeRoleWithSAMLResponse extends $tea.Model {
    constructor(map) {
        super(map);
    }
    static names() {
        return {
            headers: 'headers',
            statusCode: 'statusCode',
            body: 'body',
        };
    }
    static types() {
        return {
            headers: { 'type': 'map', 'keyType': 'string', 'valueType': 'string' },
            statusCode: 'number',
            body: AssumeRoleWithSAMLResponseBody,
        };
    }
}
exports.AssumeRoleWithSAMLResponse = AssumeRoleWithSAMLResponse;
class GetCallerIdentityResponseBody extends $tea.Model {
    constructor(map) {
        super(map);
    }
    static names() {
        return {
            accountId: 'AccountId',
            arn: 'Arn',
            identityType: 'IdentityType',
            principalId: 'PrincipalId',
            requestId: 'RequestId',
            roleId: 'RoleId',
            userId: 'UserId',
        };
    }
    static types() {
        return {
            accountId: 'string',
            arn: 'string',
            identityType: 'string',
            principalId: 'string',
            requestId: 'string',
            roleId: 'string',
            userId: 'string',
        };
    }
}
exports.GetCallerIdentityResponseBody = GetCallerIdentityResponseBody;
class GetCallerIdentityResponse extends $tea.Model {
    constructor(map) {
        super(map);
    }
    static names() {
        return {
            headers: 'headers',
            statusCode: 'statusCode',
            body: 'body',
        };
    }
    static types() {
        return {
            headers: { 'type': 'map', 'keyType': 'string', 'valueType': 'string' },
            statusCode: 'number',
            body: GetCallerIdentityResponseBody,
        };
    }
}
exports.GetCallerIdentityResponse = GetCallerIdentityResponse;
class AssumeRoleResponseBodyAssumedRoleUser extends $tea.Model {
    constructor(map) {
        super(map);
    }
    static names() {
        return {
            arn: 'Arn',
            assumedRoleId: 'AssumedRoleId',
        };
    }
    static types() {
        return {
            arn: 'string',
            assumedRoleId: 'string',
        };
    }
}
exports.AssumeRoleResponseBodyAssumedRoleUser = AssumeRoleResponseBodyAssumedRoleUser;
class AssumeRoleResponseBodyCredentials extends $tea.Model {
    constructor(map) {
        super(map);
    }
    static names() {
        return {
            accessKeyId: 'AccessKeyId',
            accessKeySecret: 'AccessKeySecret',
            expiration: 'Expiration',
            securityToken: 'SecurityToken',
        };
    }
    static types() {
        return {
            accessKeyId: 'string',
            accessKeySecret: 'string',
            expiration: 'string',
            securityToken: 'string',
        };
    }
}
exports.AssumeRoleResponseBodyCredentials = AssumeRoleResponseBodyCredentials;
class AssumeRoleWithOIDCResponseBodyAssumedRoleUser extends $tea.Model {
    constructor(map) {
        super(map);
    }
    static names() {
        return {
            arn: 'Arn',
            assumedRoleId: 'AssumedRoleId',
        };
    }
    static types() {
        return {
            arn: 'string',
            assumedRoleId: 'string',
        };
    }
}
exports.AssumeRoleWithOIDCResponseBodyAssumedRoleUser = AssumeRoleWithOIDCResponseBodyAssumedRoleUser;
class AssumeRoleWithOIDCResponseBodyCredentials extends $tea.Model {
    constructor(map) {
        super(map);
    }
    static names() {
        return {
            accessKeyId: 'AccessKeyId',
            accessKeySecret: 'AccessKeySecret',
            expiration: 'Expiration',
            securityToken: 'SecurityToken',
        };
    }
    static types() {
        return {
            accessKeyId: 'string',
            accessKeySecret: 'string',
            expiration: 'string',
            securityToken: 'string',
        };
    }
}
exports.AssumeRoleWithOIDCResponseBodyCredentials = AssumeRoleWithOIDCResponseBodyCredentials;
class AssumeRoleWithOIDCResponseBodyOIDCTokenInfo extends $tea.Model {
    constructor(map) {
        super(map);
    }
    static names() {
        return {
            clientIds: 'ClientIds',
            expirationTime: 'ExpirationTime',
            issuanceTime: 'IssuanceTime',
            issuer: 'Issuer',
            subject: 'Subject',
            verificationInfo: 'VerificationInfo',
        };
    }
    static types() {
        return {
            clientIds: 'string',
            expirationTime: 'string',
            issuanceTime: 'string',
            issuer: 'string',
            subject: 'string',
            verificationInfo: 'string',
        };
    }
}
exports.AssumeRoleWithOIDCResponseBodyOIDCTokenInfo = AssumeRoleWithOIDCResponseBodyOIDCTokenInfo;
class AssumeRoleWithSAMLResponseBodyAssumedRoleUser extends $tea.Model {
    constructor(map) {
        super(map);
    }
    static names() {
        return {
            arn: 'Arn',
            assumedRoleId: 'AssumedRoleId',
        };
    }
    static types() {
        return {
            arn: 'string',
            assumedRoleId: 'string',
        };
    }
}
exports.AssumeRoleWithSAMLResponseBodyAssumedRoleUser = AssumeRoleWithSAMLResponseBodyAssumedRoleUser;
class AssumeRoleWithSAMLResponseBodyCredentials extends $tea.Model {
    constructor(map) {
        super(map);
    }
    static names() {
        return {
            accessKeyId: 'AccessKeyId',
            accessKeySecret: 'AccessKeySecret',
            expiration: 'Expiration',
            securityToken: 'SecurityToken',
        };
    }
    static types() {
        return {
            accessKeyId: 'string',
            accessKeySecret: 'string',
            expiration: 'string',
            securityToken: 'string',
        };
    }
}
exports.AssumeRoleWithSAMLResponseBodyCredentials = AssumeRoleWithSAMLResponseBodyCredentials;
class AssumeRoleWithSAMLResponseBodySAMLAssertionInfo extends $tea.Model {
    constructor(map) {
        super(map);
    }
    static names() {
        return {
            issuer: 'Issuer',
            recipient: 'Recipient',
            subject: 'Subject',
            subjectType: 'SubjectType',
        };
    }
    static types() {
        return {
            issuer: 'string',
            recipient: 'string',
            subject: 'string',
            subjectType: 'string',
        };
    }
}
exports.AssumeRoleWithSAMLResponseBodySAMLAssertionInfo = AssumeRoleWithSAMLResponseBodySAMLAssertionInfo;
class Client extends openapi_client_1.default {
    constructor(config) {
        super(config);
        this._signatureAlgorithm = "v2";
        this._endpointRule = "regional";
        this._endpointMap = {
            'ap-northeast-2-pop': "sts.aliyuncs.com",
            'cn-beijing-finance-1': "sts.aliyuncs.com",
            'cn-beijing-finance-pop': "sts.aliyuncs.com",
            'cn-beijing-gov-1': "sts.aliyuncs.com",
            'cn-beijing-nu16-b01': "sts.aliyuncs.com",
            'cn-edge-1': "sts.aliyuncs.com",
            'cn-fujian': "sts.aliyuncs.com",
            'cn-haidian-cm12-c01': "sts.aliyuncs.com",
            'cn-hangzhou-bj-b01': "sts.aliyuncs.com",
            'cn-hangzhou-finance': "sts.aliyuncs.com",
            'cn-hangzhou-internal-prod-1': "sts.aliyuncs.com",
            'cn-hangzhou-internal-test-1': "sts.aliyuncs.com",
            'cn-hangzhou-internal-test-2': "sts.aliyuncs.com",
            'cn-hangzhou-internal-test-3': "sts.aliyuncs.com",
            'cn-hangzhou-test-306': "sts.aliyuncs.com",
            'cn-hongkong-finance-pop': "sts.aliyuncs.com",
            'cn-huhehaote-nebula-1': "sts.aliyuncs.com",
            'cn-north-2-gov-1': "sts-vpc.cn-north-2-gov-1.aliyuncs.com",
            'cn-qingdao-nebula': "sts.aliyuncs.com",
            'cn-shanghai-et15-b01': "sts.aliyuncs.com",
            'cn-shanghai-et2-b01': "sts.aliyuncs.com",
            'cn-shanghai-inner': "sts.aliyuncs.com",
            'cn-shanghai-internal-test-1': "sts.aliyuncs.com",
            'cn-shenzhen-finance-1': "sts-vpc.cn-shenzhen-finance-1.aliyuncs.com",
            'cn-shenzhen-inner': "sts.aliyuncs.com",
            'cn-shenzhen-st4-d01': "sts.aliyuncs.com",
            'cn-shenzhen-su18-b01': "sts.aliyuncs.com",
            'cn-wuhan': "sts.aliyuncs.com",
            'cn-yushanfang': "sts.aliyuncs.com",
            'cn-zhangbei': "sts.aliyuncs.com",
            'cn-zhangbei-na61-b01': "sts.aliyuncs.com",
            'cn-zhangjiakou-na62-a01': "sts.aliyuncs.com",
            'cn-zhengzhou-nebula-1': "sts.aliyuncs.com",
            'eu-west-1-oxs': "sts.aliyuncs.com",
            'rus-west-1-pop': "sts.aliyuncs.com",
        };
        this.checkConfig(config);
        this._endpoint = this.getEndpoint("sts", this._regionId, this._endpointRule, this._network, this._suffix, this._endpointMap, this._endpoint);
    }
    getEndpoint(productId, regionId, endpointRule, network, suffix, endpointMap, endpoint) {
        if (!tea_util_1.default.empty(endpoint)) {
            return endpoint;
        }
        if (!tea_util_1.default.isUnset(endpointMap) && !tea_util_1.default.empty(endpointMap[regionId])) {
            return endpointMap[regionId];
        }
        return endpoint_util_1.default.getEndpointRules(productId, regionId, endpointRule, network, suffix);
    }
    /**
      * ### Prerequisites
      * You cannot use an Alibaba Cloud account to call this operation. The requester of this operation can only be a RAM user or RAM role. Make sure that the AliyunSTSAssumeRoleAccess policy is attached to the requester. After this policy is attached to the requester, the requester has the management permissions on STS.
      * If you do not attach the AliyunSTSAssumeRoleAccess policy to the requester, the following error message is returned:
      * `You are not authorized to do this action. You should be authorized by RAM.`
      * You can refer to the following information to troubleshoot the error:
      * *   Cause of the error: The policy that is required to assume a RAM role is not attached to the requester. To resolve this issue, attach the AliyunSTSAssumeRoleAccess policy or a custom policy to the requester. For more information, see [Can I specify the RAM role that a RAM user can assume?](~~39744~~) and [Grant permissions to a RAM user](~~116146~~).
      * *   Cause of the error: The requester is not authorized to assume the RAM role. To resolve this issue, add the requester to the Principal element in the trust policy of the RAM role For more information, see [Edit the trust policy of a RAM role](~~116819~~).
      * ### Best practices
      * An STS token is valid for a period of time after it is issued, and the number of STS tokens that can be issued within an interval is also limited. Therefore, we recommend that you configure a proper validity period for an STS token and repeatedly use the token within this period. This prevents frequent issuing of STS tokens from adversely affecting your services if a large number of requests are sent. For more information about the limit, see [Is the number of STS API requests limited?](~~39744~~) You can configure the `DurationSeconds` parameter to specify a validity period for an STS token.
      * When you upload or download Object Storage Service (OSS) objects on mobile devices, a large number of STS API requests are sent. In this case, repeated use of an STS token may not meet your business requirements. To avoid the limit on STS API requests from affecting access to OSS, you can **add a signature to the URL of an OSS object**. For more information, see [Add signatures to URLs](~~31952~~) and [Obtain signature information from the server and upload data to OSS](~~31926~~).
      *
      * @param request AssumeRoleRequest
      * @param runtime runtime options for this request RuntimeOptions
      * @return AssumeRoleResponse
     */
    async assumeRoleWithOptions(request, runtime) {
        tea_util_1.default.validateModel(request);
        let query = {};
        if (!tea_util_1.default.isUnset(request.durationSeconds)) {
            query["DurationSeconds"] = request.durationSeconds;
        }
        if (!tea_util_1.default.isUnset(request.externalId)) {
            query["ExternalId"] = request.externalId;
        }
        if (!tea_util_1.default.isUnset(request.policy)) {
            query["Policy"] = request.policy;
        }
        if (!tea_util_1.default.isUnset(request.roleArn)) {
            query["RoleArn"] = request.roleArn;
        }
        if (!tea_util_1.default.isUnset(request.roleSessionName)) {
            query["RoleSessionName"] = request.roleSessionName;
        }
        let req = new $OpenApi.OpenApiRequest({
            query: openapi_util_1.default.query(query),
        });
        let params = new $OpenApi.Params({
            action: "AssumeRole",
            version: "2015-04-01",
            protocol: "HTTPS",
            pathname: "/",
            method: "POST",
            authType: "AK",
            style: "RPC",
            reqBodyType: "formData",
            bodyType: "json",
        });
        return $tea.cast(await this.callApi(params, req, runtime), new AssumeRoleResponse({}));
    }
    /**
      * ### Prerequisites
      * You cannot use an Alibaba Cloud account to call this operation. The requester of this operation can only be a RAM user or RAM role. Make sure that the AliyunSTSAssumeRoleAccess policy is attached to the requester. After this policy is attached to the requester, the requester has the management permissions on STS.
      * If you do not attach the AliyunSTSAssumeRoleAccess policy to the requester, the following error message is returned:
      * `You are not authorized to do this action. You should be authorized by RAM.`
      * You can refer to the following information to troubleshoot the error:
      * *   Cause of the error: The policy that is required to assume a RAM role is not attached to the requester. To resolve this issue, attach the AliyunSTSAssumeRoleAccess policy or a custom policy to the requester. For more information, see [Can I specify the RAM role that a RAM user can assume?](~~39744~~) and [Grant permissions to a RAM user](~~116146~~).
      * *   Cause of the error: The requester is not authorized to assume the RAM role. To resolve this issue, add the requester to the Principal element in the trust policy of the RAM role For more information, see [Edit the trust policy of a RAM role](~~116819~~).
      * ### Best practices
      * An STS token is valid for a period of time after it is issued, and the number of STS tokens that can be issued within an interval is also limited. Therefore, we recommend that you configure a proper validity period for an STS token and repeatedly use the token within this period. This prevents frequent issuing of STS tokens from adversely affecting your services if a large number of requests are sent. For more information about the limit, see [Is the number of STS API requests limited?](~~39744~~) You can configure the `DurationSeconds` parameter to specify a validity period for an STS token.
      * When you upload or download Object Storage Service (OSS) objects on mobile devices, a large number of STS API requests are sent. In this case, repeated use of an STS token may not meet your business requirements. To avoid the limit on STS API requests from affecting access to OSS, you can **add a signature to the URL of an OSS object**. For more information, see [Add signatures to URLs](~~31952~~) and [Obtain signature information from the server and upload data to OSS](~~31926~~).
      *
      * @param request AssumeRoleRequest
      * @return AssumeRoleResponse
     */
    async assumeRole(request) {
        let runtime = new $Util.RuntimeOptions({});
        return await this.assumeRoleWithOptions(request, runtime);
    }
    /**
      * ### Prerequisites
      * *   An OIDC token is obtained from an external identity provider (IdP).
      * *   An OIDC IdP is created in the RAM console. For more information, see [Create an OIDC IdP](~~327123~~) or [CreateOIDCProvider](~~327135~~).
      * *   A RAM role whose trusted entity is an OIDC IdP is created in the RAM console. For more information, see [Create a RAM role for a trusted IdP](~~116805~~) or [CreateRole](~~28710~~).
      *
      * @param request AssumeRoleWithOIDCRequest
      * @param runtime runtime options for this request RuntimeOptions
      * @return AssumeRoleWithOIDCResponse
     */
    async assumeRoleWithOIDCWithOptions(request, runtime) {
        tea_util_1.default.validateModel(request);
        let query = {};
        if (!tea_util_1.default.isUnset(request.durationSeconds)) {
            query["DurationSeconds"] = request.durationSeconds;
        }
        if (!tea_util_1.default.isUnset(request.OIDCProviderArn)) {
            query["OIDCProviderArn"] = request.OIDCProviderArn;
        }
        if (!tea_util_1.default.isUnset(request.OIDCToken)) {
            query["OIDCToken"] = request.OIDCToken;
        }
        if (!tea_util_1.default.isUnset(request.policy)) {
            query["Policy"] = request.policy;
        }
        if (!tea_util_1.default.isUnset(request.roleArn)) {
            query["RoleArn"] = request.roleArn;
        }
        if (!tea_util_1.default.isUnset(request.roleSessionName)) {
            query["RoleSessionName"] = request.roleSessionName;
        }
        let req = new $OpenApi.OpenApiRequest({
            query: openapi_util_1.default.query(query),
        });
        let params = new $OpenApi.Params({
            action: "AssumeRoleWithOIDC",
            version: "2015-04-01",
            protocol: "HTTPS",
            pathname: "/",
            method: "POST",
            authType: "Anonymous",
            style: "RPC",
            reqBodyType: "formData",
            bodyType: "json",
        });
        return $tea.cast(await this.callApi(params, req, runtime), new AssumeRoleWithOIDCResponse({}));
    }
    /**
      * ### Prerequisites
      * *   An OIDC token is obtained from an external identity provider (IdP).
      * *   An OIDC IdP is created in the RAM console. For more information, see [Create an OIDC IdP](~~327123~~) or [CreateOIDCProvider](~~327135~~).
      * *   A RAM role whose trusted entity is an OIDC IdP is created in the RAM console. For more information, see [Create a RAM role for a trusted IdP](~~116805~~) or [CreateRole](~~28710~~).
      *
      * @param request AssumeRoleWithOIDCRequest
      * @return AssumeRoleWithOIDCResponse
     */
    async assumeRoleWithOIDC(request) {
        let runtime = new $Util.RuntimeOptions({});
        return await this.assumeRoleWithOIDCWithOptions(request, runtime);
    }
    /**
      * ###
      * *   A SAML response is obtained from an external identity provider (IdP).
      * *   A SAML IdP is created in the RAM console. For more information, see [Create a SAML IdP](~~116083~~) or [CreateSAMLProvider](~~186846~~).
      * *   A RAM role whose trusted entity is a SAML IdP is created in the RAM console. For more information, see [Create a RAM role for a trusted IdP](~~116805~~) or [CreateRole](~~28710~~).
      *
      * @param request AssumeRoleWithSAMLRequest
      * @param runtime runtime options for this request RuntimeOptions
      * @return AssumeRoleWithSAMLResponse
     */
    async assumeRoleWithSAMLWithOptions(request, runtime) {
        tea_util_1.default.validateModel(request);
        let query = {};
        if (!tea_util_1.default.isUnset(request.durationSeconds)) {
            query["DurationSeconds"] = request.durationSeconds;
        }
        if (!tea_util_1.default.isUnset(request.policy)) {
            query["Policy"] = request.policy;
        }
        if (!tea_util_1.default.isUnset(request.roleArn)) {
            query["RoleArn"] = request.roleArn;
        }
        if (!tea_util_1.default.isUnset(request.SAMLAssertion)) {
            query["SAMLAssertion"] = request.SAMLAssertion;
        }
        if (!tea_util_1.default.isUnset(request.SAMLProviderArn)) {
            query["SAMLProviderArn"] = request.SAMLProviderArn;
        }
        let req = new $OpenApi.OpenApiRequest({
            query: openapi_util_1.default.query(query),
        });
        let params = new $OpenApi.Params({
            action: "AssumeRoleWithSAML",
            version: "2015-04-01",
            protocol: "HTTPS",
            pathname: "/",
            method: "POST",
            authType: "Anonymous",
            style: "RPC",
            reqBodyType: "formData",
            bodyType: "json",
        });
        return $tea.cast(await this.callApi(params, req, runtime), new AssumeRoleWithSAMLResponse({}));
    }
    /**
      * ###
      * *   A SAML response is obtained from an external identity provider (IdP).
      * *   A SAML IdP is created in the RAM console. For more information, see [Create a SAML IdP](~~116083~~) or [CreateSAMLProvider](~~186846~~).
      * *   A RAM role whose trusted entity is a SAML IdP is created in the RAM console. For more information, see [Create a RAM role for a trusted IdP](~~116805~~) or [CreateRole](~~28710~~).
      *
      * @param request AssumeRoleWithSAMLRequest
      * @return AssumeRoleWithSAMLResponse
     */
    async assumeRoleWithSAML(request) {
        let runtime = new $Util.RuntimeOptions({});
        return await this.assumeRoleWithSAMLWithOptions(request, runtime);
    }
    async getCallerIdentityWithOptions(runtime) {
        let req = new $OpenApi.OpenApiRequest({});
        let params = new $OpenApi.Params({
            action: "GetCallerIdentity",
            version: "2015-04-01",
            protocol: "HTTPS",
            pathname: "/",
            method: "POST",
            authType: "AK",
            style: "RPC",
            reqBodyType: "formData",
            bodyType: "json",
        });
        return $tea.cast(await this.callApi(params, req, runtime), new GetCallerIdentityResponse({}));
    }
    async getCallerIdentity() {
        let runtime = new $Util.RuntimeOptions({});
        return await this.getCallerIdentityWithOptions(runtime);
    }
}
exports.default = Client;
